notebooks/linear_regression.livemd

```diff
@@ -0,0 +1,610 @@
+
+Firstly, we generate the data which simulates the function $f(x) = 3x + 4$ with added uniform, zero-mean noise. `Nx.Random.uniform` creates a tensor with a given shape and type.
+
+```elixir
+key = Nx.Random.key(42)
+size = 100
+{uniform_x, new_key} = Nx.Random.uniform(key, shape: {size, 1}, type: :f32)
+{uniform_y, _} = Nx.Random.uniform(new_key, shape: {size, 1}, type: :f32)
+x = 2 * uniform_x
+y = 3 * x + 4 + (uniform_y - Nx.mean(uniform_y))
+```
+
+Now let's plot the generated points.
+
+```elixir
+df = Explorer.DataFrame.new(x: Nx.to_flat_list(x), y: Nx.to_flat_list(y))
+
+Vl.new(
+  title: [
+    text: "Scatterplot of Generated Data",
+    offset: 20
+  ],
+  width: 500,
+  height: 500
+)
+|> Vl.data_from_values(df)
+|> Vl.mark(:circle)
+|> Vl.encode_field(:x, "x",
+  type: :quantitative,
+  scale: [domain: [-0.05, 2.05]],
+  axis: [grid: false]
+)
+|> Vl.encode_field(:y, "y", type: :quantitative, scale: [domain: [2.5, 12]], axis: [grid: false])
+```
+
+For a regression task, we will use the `Scholar.Linear.LinearRegression` module.
+
+```elixir
+model = Scholar.Linear.LinearRegression.fit(x, y)
+```
+
+As we can see, the coefficient is almost 3.0, and the intercept is nearly 4.0. Those are decent estimations. They are not exactly equal to 3.0 and 4.0 because we introduce noise to our samples.
+
+Now, let's plot the result of linear regression.
+
+```elixir
+intercept = model.intercept |> Nx.reshape({}) |> Nx.to_number()
+coefficients = model.coefficients |> Nx.reshape({}) |> Nx.to_number()
+x_1 = 0
+x_2 = 2
+
+df_results =
+  Explorer.DataFrame.new(
+    x: [x_1, x_2],
+    y: [
+      x_1 * coefficients + intercept,
+      x_2 * coefficients + intercept
+    ]
+  )
+
+Vl.new(
+  title: [
+    text: "Scatterplot of Generated Data",
+    offset: 20
+  ],
+  width: 500,
+  height: 500
+)
+|> Vl.layers([
+  Vl.new()
+  |> Vl.data_from_values(df)
+  |> Vl.mark(:circle)
+  |> Vl.encode_field(:x, "x",
+    type: :quantitative,
+    scale: [domain: [-0.05, 2.05]],
+    axis: [grid: false]
+  )
+  |> Vl.encode_field(:y, "y", type: :quantitative, scale: [domain: [2.5, 12]], axis: [grid: false]),
+  Vl.new()
+  |> Vl.data_from_values(df_results)
+  |> Vl.mark(:line, color: :green)
+  |> Vl.encode_field(:x, "x", type: :quantitative, scale: [domain: [-0.05, 2.05]])
+  |> Vl.encode_field(:y, "y", type: :quantitative, scale: [domain: [2.5, 12]])
+])
+```
+
+Using `Scholar.Linear.LinearRegression.predict`, we can predict an expected value for a given input. However, we must remember that our prediction will be valid only if we consider linearly dependent data. Fortunately, our data set is perfect for this kind of prediction.
+
+Now we will predict one value and draw it on the previous graph in a different color.
+
+```elixir
+x_prediction = Nx.tensor([[0.83]])
+
+y_prediction =
+  Scholar.Linear.LinearRegression.predict(model, x_prediction) |> Nx.reshape({}) |> Nx.to_number()
+
+x_prediction = x_prediction |> Nx.reshape({}) |> Nx.to_number()
```

Comment:
```suggestion
x_prediction = Nx.tensor([[0.83]])  |> Nx.reshape({}) |> Nx.to_number()

y_prediction =
  Scholar.Linear.LinearRegression.predict(model, x_prediction)
```

Reply:
If this doesn't work, we should change LinearRegression.predict to accept as a valid input. I see no reason for it to force a 2D tensor as an input. Output should have the same shape as the input, though.

Reply:
I think that this might have a reason. For training, we pass 2D data which has `n_samples` and `n_features`. So it somehow forces input for prediction to have the same `n_features` which is the first axis. But features are considered usually as the second axis.  For scalar tensors, we can guess that input has one row and one feature (but we virtually add those axes). For one-dimensional input, it will be ambiguous if the input has one row with n features or n rows with 1 feature. There is no specific reason to assume one version. If we wrap it into a 2D tensor then it is clear what we mean.

Reply:
I believe that's why sklearn also forces users to use only 2D input.

Reply:
I don't think it would be ambiguous if we define/standardize on 1D tensors being a single example (i.e. 1 row, n features)

Reply:
Ok, I will introduce such feature in a separate PR and then update the livebook.

Reply:
Resolving this in favor of the PR